File: core/src/main/java/com/mcreater/amclcore/account/AbstractAccount.java
package com.mcreater.amclcore.account;

import com.mcreater.amclcore.concurrent.task.AbstractAction;
import com.mcreater.amclcore.concurrent.task.AbstractTask;
import com.mcreater.amclcore.model.oauth.session.MinecraftNameChangedTimeRequestModel;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.UUID;

import static java.util.Objects.requireNonNull;

public abstract class AbstractAccount {
    public enum UserType {
        MOJANG,
        OFFLINE
    }

    @Getter
    private String accountName;
    @Getter
    private UUID uuid;
    @Getter
    private String accessToken;

    protected void setAccountName(@NotNull String accountName) {
        this.accountName = requireNonNull(accountName);
    }

    protected void setUuid(@NotNull UUID uuid) {
        this.uuid = requireNonNull(uuid);
    }

    protected void setAccessToken(@NotNull String accessToken) {
        this.accessToken = requireNonNull(accessToken);
    }

    public AbstractAccount(@NotNull String accountName, @NotNull UUID uuid, @NotNull String accessToken) {
        setAccountName(accountName);
        setUuid(uuid);
        setAccessToken(accessToken);
    }

    public AbstractAccount(@NotNull String accessToken) {
        setAccessToken(accessToken);
    }

    public AbstractAccount(@NotNull String accountName, @NotNull UUID uuid) {
        setAccountName(accountName);
        setUuid(uuid);
    }

    /**
     * create account refresh task<br>
     * 创建账户刷新任务
     *
     * @return the created task<br>被创建的任务
     */
    public abstract AbstractAction refreshAsync();

    /**
     * create fetch profile task<br>
     * 创建档案获取任务
     *
     * @return the created task<br>被创建的任务
     */
    public abstract AbstractAction fetchProfileAsync();

    /**
     * create check profile's AccessToken is outdated or not task<br>
     * 创建检查账户 AccessToken 是否失效与否的任务
     *
     * @return the created task<br>被创建的任务
     */
    public abstract AbstractTask<Boolean> validateAccountAsync();

    /**
     * create disable cape task<br>
     * 创建禁用披风任务
     *
     * @return the created task<br>被创建的任务
     */
    public abstract AbstractAction disableAccountCapeAsync();

    /**
     * create enable selected cape task<br>
     * 创建启用选定披风任务
     *
     * @param id the cape index<br>披风序列号
     * @return the created task<br>被创建的任务
     */
    public abstract AbstractAction enableAccountCapeAsync(String id);

    /**
     * create check account name changeable task<br>
     * 创建检查账户名是否可以改变的任务
     *
     * @return the created task<br>被创建的任务
     */
    public abstract AbstractTask<Boolean> checkAccountNameChangeableAsync(@NotNull String newName);

    /**
     * create change account name task<br>
     * 创建更改账户名任务
     *
     * @return the created task<br>被创建的任务
     */
    public abstract AbstractAction changeAccountNameAsync(@NotNull String newName);

    /**
     * create check account name changed time task<br>
     * 创建检查账户名改变时间的任务
     *
     * @return the created task<br>被创建的任务
     */
    public abstract AbstractTask<MinecraftNameChangedTimeRequestModel> checkAccountNameChangedTimeAsync();

    /**
     * check account name is allowed<br>
     * 检查账户名是否允许
     *
     * @return the check result<br>检查结果
     */
    public abstract boolean accountNameAllowed(String name);

    /**
     * create reset skin task<br>
     * 创建重置皮肤任务
     *
     * @return the created task<br>被创建的任务
     */
    public abstract AbstractAction resetSkinAsync();

    /**
     * create upload skin task<br>
     * 创建上传皮肤任务
     *
     * @return the created task<br>被创建的任务
     */
    public abstract AbstractAction uploadSkinAsync(File file, boolean isSlim);

    public final MicrosoftAccount toMicrosoftAccount() {
        return (MicrosoftAccount) this;
    }

    public final OfflineAccount toOffLineAccount() {
        return (OfflineAccount) this;
    }

    public final boolean isMicrosoftAccount() {
        return this instanceof MicrosoftAccount;
    }

    public final boolean isOffLineAccount() {
        return this instanceof OfflineAccount;
    }
}

File: core/src/main/java/com/mcreater/amclcore/account/AbstractAccount.kt
package com.mcreater.amclcore.account

import com.mcreater.amclcore.command.CommandArg
import com.mcreater.amclcore.concurrent.task.AbstractAction
import com.mcreater.amclcore.concurrent.task.AbstractTask
import com.mcreater.amclcore.model.oauth.session.MinecraftNameChangedTimeRequestModel
import com.mcreater.amclcore.util.StringUtil
import java.io.File
import java.util.*


abstract class AbstractAccount {
    enum class UserType {
        MOJANG,
        OFFLINE
    }

    open var accountName: String? = null
    open var uuid: UUID? = null
    var accessToken: String? = null

    constructor(accountName: String, uuid: UUID, accessToken: String) {
        this.accountName = accountName
        this.uuid = uuid
        this.accessToken = accessToken
    }

    constructor(accessToken: String) {
        this.accessToken = accessToken
    }

    constructor(accountName: String, uuid: UUID) {
        this.accountName = accountName
        this.uuid = uuid
    }

    /**
     * create account refresh task<br></br>
     * 创建账户刷新任务
     *
     * @return the created task<br></br>被创建的任务
     */
    abstract fun refreshAsync(): AbstractAction?

    /**
     * create fetch profile task<br></br>
     * 创建档案获取任务
     *
     * @return the created task<br></br>被创建的任务
     */
    abstract fun fetchProfileAsync(): AbstractAction?

    /**
     * create check profile's AccessToken is outdated or not task<br></br>
     * 创建检查账户 AccessToken 是否失效与否的任务
     *
     * @return the created task<br></br>被创建的任务
     */
    abstract fun validateAccountAsync(): AbstractTask<Boolean?>?

    /**
     * create disable cape task<br></br>
     * 创建禁用披风任务
     *
     * @return the created task<br></br>被创建的任务
     */
    abstract fun disableAccountCapeAsync(): AbstractAction?

    /**
     * create enable selected cape task<br></br>
     * 创建启用选定披风任务
     *
     * @param id the cape index<br></br>披风序列号
     * @return the created task<br></br>被创建的任务
     */
    abstract fun enableAccountCapeAsync(id: String?): AbstractAction?

    /**
     * create check account name changeable task<br></br>
     * 创建检查账户名是否可以改变的任务
     *
     * @return the created task<br></br>被创建的任务
     */
    abstract fun checkAccountNameChangeableAsync(newName: String): AbstractTask<Boolean?>?

    /**
     * create change account name task<br></br>
     * 创建更改账户名任务
     *
     * @return the created task<br></br>被创建的任务
     */
    abstract fun changeAccountNameAsync(newName: String): AbstractAction?

    /**
     * create check account name changed time task<br></br>
     * 创建检查账户名改变时间的任务
     *
     * @return the created task<br></br>被创建的任务
     */
    abstract fun checkAccountNameChangedTimeAsync(): AbstractTask<MinecraftNameChangedTimeRequestModel?>?

    /**
     * check account name is allowed<br></br>
     * 检查账户名是否允许
     *
     * @return the check result<br></br>检查结果
     */
    abstract fun accountNameAllowed(name: String?): Boolean

    /**
     * create reset skin task<br></br>
     * 创建重置皮肤任务
     *
     * @return the created task<br></br>被创建的任务
     */
    abstract fun resetSkinAsync(): AbstractAction?

    /**
     * create upload skin task<br></br>
     * 创建上传皮肤任务
     *
     * @return the created task<br></br>被创建的任务
     */
    abstract fun uploadSkinAsync(file: File?, isSlim: Boolean): AbstractAction?
    fun toMicrosoftAccount(): MicrosoftAccount {
        return this as MicrosoftAccount
    }

    fun toOffLineAccount(): OfflineAccount {
        return this as OfflineAccount
    }

    val isMicrosoftAccount: Boolean
        get() = this is MicrosoftAccount
    val isOffLineAccount: Boolean
        get() = this is OfflineAccount
    abstract val addonArgs: List<CommandArg?>?

    abstract fun preLaunchAsync(): AbstractAction?
    fun toProfile(): Profile {
        return Profile(StringUtil.toNoLineUUID(uuid), accountName)
    }

    data class Profile(
        val id: String? = null,
        val name: String? = null
    )
}
